src/miner.cpp

```diff
@@ -649,6 +709,93 @@ void BlockAssembler::SortForBlock(const CTxMemPool::setEntries& package, std::ve
     std::sort(sortedEntries.begin(), sortedEntries.end(), CompareTxIterByEntryTime());
 }
 
+bool BlockAssembler::EvmTxPreapply(const EvmTxPreApplyContext& ctx) {
+    auto txMessage = customTypeToMessage(ctx.txType);
+    auto& txIter = ctx.txIter;
+    auto& evmQueueId = ctx.evmQueueId;
+    auto& metadata = ctx.txMetadata;
+    auto& height = ctx.height;
+    auto& pkgCtx = ctx.pkgCtx;
+    auto& checkedDfTxHashSet = ctx.checkedTXEntries;
+    auto& failedTxSet = ctx.failedTxEntries;
+    auto& failedNonces = ctx.pkgCtx.failedNonces;
+    auto& replaceByFee = ctx.pkgCtx.replaceByFee;
+
+    auto txIters = [](std::map<uint64_t, CTxMemPool::txiter> &iterMap) -> std::vector<CTxMemPool::txiter> {
+        std::vector<CTxMemPool::txiter> txIters;
+        for (const auto& [nonce, it] : iterMap) {
+            txIters.push_back(it);
+        }
+        return txIters;
+    };
+
+    if (!CustomMetadataParse(height, Params().GetConsensus(), metadata, txMessage)) {
+        return false;
+    }
+
+    const auto obj = std::get<CEvmTxMessage>(txMessage);
+
+    CrossBoundaryResult result;
+    const auto txResult = evm_try_prevalidate_raw_tx(result, HexStr(obj.evmTx));
+    if (!result.ok) {
+        return false;
+    }
+
+    auto& evmFeeMap = pkgCtx.feeMap;
+    auto& evmAddressTxsMap = pkgCtx.addressTxsMap;
+
+    const auto addrKey = EvmAddressWithNonce { txResult.sender, txResult.nonce }; 
+    if (auto feeEntry = evmFeeMap.find(addrKey); feeEntry != evmFeeMap.end()) {
+        // Key already exists. We check to see if we need to prioritize higher fee tx
+        const auto& lastFee = feeEntry->second;
+        if (txResult.prepay_fee > lastFee) {
+            // Higher paying fee. Remove all TXs from sender and add to collection to add them again in order.
+            auto addrTxs = evmAddressTxsMap[addrKey.address];
+            for (auto iter: txIters(addrTxs)) {
+                if (RemoveFromBlock(iter)) {
+                    inBlock.erase(iter);
+                }
+            }
+            // Remove all fee entries relating to the address
+            for (auto it = evmFeeMap.begin(); it != evmFeeMap.end();) {
+                const auto& [sender, nonce] = it->first;
+                if (sender == addrKey.address) {
+                    it = evmFeeMap.erase(it);
+                } else {
+                    ++it;
+                }
+            }
+            // Buggy code to fix below:
+            checkedDfTxHashSet.erase(addrTxs[addrKey.nonce]->GetTx().GetHash());
+            addrTxs[addrKey.nonce] = txIter;
+            auto count{addrKey.nonce};
+            for (const auto& [nonce, entry] : addrTxs) {
+                checkedDfTxHashSet.erase(entry->GetTx().GetHash());
+                replaceByFee.emplace(count, entry);
+                ++count;
+            }
+            evmAddressTxsMap.erase(addrKey.address);
+            evm_remove_txs_by_sender(evmQueueId, addrKey.address);
```

Comment:
Same thing as caught in PR #2220. When we call evm_remove_txs_by_sender, the current pipeline is such that we are actually removing all transactions from that specific sender in the txqueue. This actually means that if for example we send 5 evm txs from 0 to 5 nonces, and then an evm tx of nonce 3 with higher prepay gas is submitted which replaces the evm tx nonce 3 - we are wiping all txs from 0 to 5 here. This test isnt actually covered in feature_evm, since the nonce replacement is nicely replacing nonce 0 in the test case which misses this specific edge case.

Reply:
Resolving as the current pipeline is correct. We are wiping all of our indexes whenever we replace the evm tx (both ancestors and descendants), and adding them into the replacebyfee index.